### PR TITLE
SITL precland

### PR DESCRIPTION
**Pull request by khancyr**
add better simulation for precland.
the precland beacon is now movable, got new diffusion radius : cylinder or cone

### CODE REVIEW

File: libraries/SITL/SIM_Precland.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    // @Description: Precland device maximun lateral range
+    // @Range: 5 100
+    // @User: Advanced
+    AP_GROUPINFO("DIST_LMT", 8, SIM_Precland, _dist_limit, 10),
+
+    AP_GROUPEND
+};
+
+/*
+  update gripper state
```
Comment (peterbarker) [inline libraries/SITL/SIM_Precland.cpp]: oops

File: libraries/SITL/SIM_Precland.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    const Location_Class origin_center(static_cast<int32_t>(_origin_lat * 1.0e7f),
+            static_cast<int32_t>(_origin_lon * 1.0e7f),
+            static_cast<int32_t>(_origin_height),
+            Location_Class::ALT_FRAME_ABOVE_HOME);
+    Vector2f center;
+    origin_center.get_vector_xy_from_origin_NE(center);
+    center = center * 0.01f;  // cm to m
+
+    if (_type == PRECLAND_TYPE_CONE) {
```
Comment (peterbarker) [inline libraries/SITL/SIM_Precland.cpp]: `switch`

File: libraries/SITL/SIM_Precland.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+#include "AP_Math/AP_Math.h"
+#include "AP_Common/Location.h"
+#include <stdio.h>
+
+using namespace SITL;
+
+// table of user settable parameters
+const AP_Param::GroupInfo SIM_Precland::var_info[] = {
+
+    // @Param: ENABLE
```
Comment (peterbarker) [inline libraries/SITL/SIM_Precland.cpp]: Should we just use the `TYPE` parameter instead with `0` meaning "none"?

Comment (khancyr) [reply libraries/SITL/SIM_Precland.cpp]: I prefer to keep ENABLE on all sitl object from a control point of view, it is simplier to enable/disable .

File: libraries/SITL/SIM_Precland.cpp (hunk 2 above)
Comment (khancyr) [reply libraries/SITL/SIM_Precland.cpp]: done

**Comment by khancyr** [comment]
correct, add sphere diffusion

File: libraries/SITL/SIM_Precland.cpp [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    // @Param: ALT_LIMIT
+    // @DisplayName: Precland device alt range
+    // @Description: Precland device maximum range altitude
+    // @Range: 0 100
+    // @User: Advanced
+    AP_GROUPINFO("ALT_LMT", 7, SIM_Precland, _alt_limit, 15),
+
+    // @Param: DIST_LIMIT
+    // @DisplayName: Precland device lateral range
+    // @Description: Precland device maximun lateral range
```
Comment (amilcarlucas) [inline libraries/SITL/SIM_Precland.cpp]: maximum

**Comment by khancyr** [comment]
rebase and correct units

**Comment by rmackay9** [comment]
.. I'll leave this one to @peterbarker to merge if that's OK.

File: libraries/SITL/SIM_Precland.cpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    // @Units: deg
+    // @Increment: 1
+    // @Range: -180 +180
+    // @User: Advanced
+    AP_GROUPINFO("YAW", 4, SIM_Precland, _orient_yaw, 0),
+
+    // @Param: RATE
+    // @DisplayName: Precland device update rate
+    // @Description: Precland device rate. e.g led patter refresh rate, RF message rate, etc.
+    // @Range: 0 15
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/SITL/SIM_Precland.cpp]: They're there now.

File: libraries/SITL/SIM_Precland.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    // @Param: TYPE
+    // @DisplayName: Precland device radiance type
+    // @Description: Precland device radiance type: it can be a cylinder, or a cone
+    // @Values: 0:cylinder,1:cone
+    // @User: Advanced
+    AP_GROUPINFO("TYPE", 6, SIM_Precland, _type, SIM_Precland::PRECLAND_TYPE_CYLINDER),
+
+    // @Param: ALT_LIMIT
+    // @DisplayName: Precland device alt range
+    // @Description: Precland device maximum range altitude
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/SITL/SIM_Precland.cpp]: They're there now.

**Comment by khancyr** [comment]
@peterbarker updated with a new IRLOCK SITL !

File: libraries/SITL/SIM_Precland.cpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    // @Description: Precland device rate. e.g led patter refresh rate, RF message rate, etc.
+    // @Units: Hz
+    // @Range: 0 200
+    // @User: Advanced
+    AP_GROUPINFO("RATE", 5, SIM_Precland, _rate, 100),
+
+    // @Param: TYPE
+    // @DisplayName: Precland device radiance type
+    // @Description: Precland device radiance type: it can be a cylinder, or a cone
+    // @Values: 0:cylinder,1:cone
```
Comment (amilcarlucas) [inline libraries/SITL/SIM_Precland.cpp]: Looks like the sphere is missing here

Comment (khancyr) [reply libraries/SITL/SIM_Precland.cpp]: done

**Comment by khancyr** [comment]
rebase and correct from comments

File: libraries/AP_IRLock/AP_IRLock_SITL_Gazebo.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+    // retrieve latest sensor data - returns true if new data is available
+    bool update() override;
+
+private:
+
+    /*
+      reply packet sent from simulator to ArduPilot
+     */
+    struct irlock_packet {
```
Comment (peterbarker) [inline libraries/AP_IRLock/AP_IRLock_SITL_Gazebo.h]: Unless you're going to pass one of these around you can move it into the sole user.

Comment (khancyr) [reply libraries/AP_IRLock/AP_IRLock_SITL_Gazebo.h]: done

**Comment by khancyr** [comment]
https://youtu.be/Ar7evwdVGqE

**Comment by rmackay9** [comment]
nice video!

**Comment by rmackay9** [comment]
A lot of effort in this PR!  I don't immediately see anything wrong..

**Comment by khancyr** [comment]
closed by https://github.com/ArduPilot/ardupilot/pull/9098/

**Pull request closed without merging**